File: parsers/indra-web_astra/src/tokens/indents/current.ts
import { Rule, Scanner, IResult } from "../../parser";
import Indent from "./indent";

export namespace CurrentIndent {
  export class Parser extends Indent.Parser {
    get defaults() {
      return {
        name: "indent-current",
        types: ["whitespace", "indent"]
      };
    }

    rule: Rule = ({ start, cursor, MATCH, UNEXPECTED }: Scanner): IResult => {
      do if (!cursor.current.char.match(/\s/)) break;
      while (cursor = cursor.next);

      if (cursor.line !== start.line
        && cursor.indents.current.length === cursor.indents.previous.length
      ) {
        return MATCH(cursor);
      } else {
        return UNEXPECTED({
          end: cursor,
          found: cursor.at(cursor.index - cursor.column).to(cursor),
          expected: cursor.indents.current.join(),
          message: "Expected the same Indentation Level"
        });
      }
    }
  }

  export const parse
    = Parser.Parse;
}

export default CurrentIndent;

File: parsers/astra-parser/src/lexer/parsers/indent/current.rs
/// A boilerplate template for a parser.
use crate::{lexer::parser, Cursor, End};

pub const KEY: &str = "indent-current";

pub struct Parser {}
impl parser::Parser for Parser {
    fn get_name(&self) -> &'static str {
        return &KEY;
    }

    fn rule(&self, cursor: &mut Cursor) -> End {
        let current_indent = cursor.indents.curr_levels();
        cursor.skip_ws();
        let next_indent = cursor.indents.curr_levels();
        if next_indent == current_indent {
            return End::Token();
        } else {
            return End::Missing(
                "level",
                &format!("indent level :{:}", current_indent),
                &format!("indent level :{:}", next_indent),
            );
        }
    }
}
